File: experiments/runs/2026-09-14_17-11-19_claim-office-example-mapping_baseline-inline-tdd-v1.1-local-git-cc_opus-5-requesty-no-thinking/src/policy.ts
import { Item, insuranceValue } from './premium.js';
import { ClaimError } from './errors.js';

export interface Damage {
  itemType: string;
  amount: number;
}

export interface ClaimResult {
  payout: number;
  remainingCap: number;
}

const DEDUCTIBLE = 100;
const CAP_FACTOR = 2;
const HALF_REIMBURSEMENT_THRESHOLD = 8;
const HALF_REIMBURSEMENT = 0.5;

/** Rounds in the MHPCO's favour: payouts go down. */
export function roundPayout(amount: number): number {
  return Math.floor(amount);
}

/**
 * Reimbursable share of a damage to the given item, before the deductible.
 *
 * The high-enchantment clause takes precedence: on an item of enchantment 8
 * or more only half the damage is reimbursed, even for dragon material. Every
 * other item — dragon material or not — is reimbursed in full.
 */
function reimbursement(item: Item, amount: number): number {
  if ((item.enchantment ?? 0) >= HALF_REIMBURSEMENT_THRESHOLD) {
    return amount * HALF_REIMBURSEMENT;
  }
  return amount;
}

export class Policy {
  readonly insuranceSum: number;
  remainingCap: number;

  constructor(private readonly items: Item[]) {
    this.insuranceSum = items.reduce(
      (sum, item) => sum + insuranceValue(item.type),
      0,
    );
    this.remainingCap = this.insuranceSum * CAP_FACTOR;
  }

  /**
   * Processes one incident. Each damage entry is a separate damage event with
   * its own deductible and is matched against one distinct insured item, so a
   * policy covering one sword cannot carry two sword damages.
   *
   * Throws without touching the cap if any entry cannot be matched or carries
   * a negative amount — the whole claim is rejected.
   */
  claim(damages: Damage[]): ClaimResult {
    const available = [...this.items];

    let total = 0;
    for (const damage of damages) {
      if (damage.amount < 0) {
        throw new ClaimError(`negative damage amount: ${damage.amount}`);
      }
      const index = available.findIndex(
        (item) => item.type === damage.itemType,
      );
      if (index === -1) {
        throw new ClaimError(
          `item not covered by the policy: ${damage.itemType}`,
        );
      }
      const [item] = available.splice(index, 1);
      total += Math.max(0, reimbursement(item, damage.amount) - DEDUCTIBLE);
    }

    const payout = roundPayout(Math.min(total, this.remainingCap));
    this.remainingCap -= payout;
    return { payout, remainingCap: this.remainingCap };
  }
}

File: experiments/runs/2026-09-21_23-56-52_claim-office-python-example-mapping_baseline-inline-tdd-v1-cc_opus-5-no-thinking/src/policy.py
"""A policy issued by a quote, and the claims settled against it."""
import math
from fractions import Fraction

import catalogue
import office
from errors import ClaimOfficeError

DEDUCTIBLE = Fraction(100)
CAP_FACTOR = 2
HALF_REIMBURSEMENT_LEVEL = 8
HALF_REIMBURSEMENT = Fraction(1, 2)
DRAGON_MATERIAL = "dragon"


def _reimbursement(item, amount):
    """The share of a damage the MHPCO reimburses, before the deductible."""
    if item.get("enchantment", 0) >= HALF_REIMBURSEMENT_LEVEL:
        return amount * HALF_REIMBURSEMENT
    return amount


class Policy:
    """Covers a fixed list of items and tracks the cap it has left."""

    def __init__(self, items):
        self._items = list(items)
        self.insurance_sum = office.insurance_sum(self._items)
        self.remaining_cap = self.insurance_sum * CAP_FACTOR

    def _matching_items(self, item_type):
        return [item for item in self._items if item["type"] == item_type]

    def _covered_item(self, damage, used):
        item_type = damage["itemType"]
        if not catalogue.is_known(item_type):
            raise ClaimOfficeError(f"unknown item type: {item_type!r}")
        covered = self._matching_items(item_type)
        index = used.get(item_type, 0)
        if index >= len(covered):
            raise ClaimOfficeError(f"{item_type!r} is not covered by this policy")
        used[item_type] = index + 1
        return covered[index]

    def settle(self, damages):
        """Pay out for one incident, consuming the policy's remaining cap."""
        used = {}
        desired = Fraction(0)
        for damage in damages:
            amount = damage["amount"]
            if amount < 0:
                raise ClaimOfficeError(f"negative damage amount: {amount}")
            item = self._covered_item(damage, used)
            reimbursed = _reimbursement(item, Fraction(amount)) - DEDUCTIBLE
            desired += max(reimbursed, Fraction(0))
        payout = math.floor(min(desired, self.remaining_cap))
        self.remaining_cap -= payout
        return payout
